File: typescript/packages/core/src/hook-runner.ts
import type { HookHandler, HookRunner } from './types/contracts.js';
import type { HookEvent } from './types/domain.js';

export class DefaultHookRunner implements HookRunner {
  private handlers = new Map<string, HookHandler>();

  register(handler: HookHandler): void {
    this.handlers.set(handler.id, handler);
  }

  async emit(event: HookEvent): Promise<void> {
    const handlers = [...this.handlers.values()].sort((a, b) => a.id.localeCompare(b.id));
    for (const handler of handlers) {
      await handler.handle(event);
    }
  }
}

File: python/src/agent_framework/core/hook_runner.py
from __future__ import annotations

from agent_framework.core.contracts import HookHandler


class DefaultHookRunner:
    def __init__(self) -> None:
        self._handlers: dict[str, HookHandler] = {}

    def register(self, handler: HookHandler) -> None:
        self._handlers[handler.id] = handler

    async def emit(self, event: dict) -> None:
        for handler in sorted(self._handlers.values(), key=lambda item: item.id):
            await handler.handle(event)
